File: 3_point-to-point_comm/solutions/3.0_comm_p2p.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

void main(int argc, char * argv[]) {
     int err;
     int nprocs, my_rank;
     err = MPI_Init(&argc, &argv);
     MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
     MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

     if (nprocs != 2 && my_rank == 0) {
       printf("ERROR: this code should be run with exactly 2 processes");
       exit(1);
     }       
     
     MPI_Status *status;
     int number;
     if (my_rank == 0) { 
       number = 0;
       printf("Before sending: my number is %d and I am proc number %d \n", number, my_rank);
       MPI_Send(&number, 1, MPI_INT, 1, 10, MPI_COMM_WORLD);
     }
     else if (my_rank == 1)  {
       number = 1;
       printf("Before receiving: my number is %d and I am proc number %d \n", number, my_rank);
       MPI_Recv(&number, 1, MPI_INT, 0, 10, MPI_COMM_WORLD, status);
     }

     printf("My number is %d and I am proc number %d \n", number, my_rank);

     err = MPI_Finalize();
}

File: 3_point-to-point_comm/solutions/3.0_comm_p2p.f90
program my_prog
   use mpi
   implicit none
   integer :: ierr, nprocs, my_rank, a
   integer status(MPI_STATUS_SIZE)
   call mpi_init(ierr)
   call mpi_comm_size(mpi_comm_world, nprocs, ierr)
   call mpi_comm_rank(mpi_comm_world, my_rank, ierr)

   if (nprocs .NE. 2 .AND. my_rank .EQ. 0) then
      write (*,*) "ERROR: this fortran code should be run with exactly two tasks"
      call exit(1)
   end if
   
   IF( my_rank .EQ. 0 ) THEN
      a = 10
      CALL MPI_SEND(a, 1, MPI_INTEGER, 1, 10, MPI_COMM_WORLD, ierr)
   ELSE IF( my_rank .EQ. 1 ) THEN
      a = 3
      CALL MPI_RECV(a, 1, MPI_INTEGER, 0, 10, MPI_COMM_WORLD, status, ierr)
   END IF
   
   write(*,*) "Hello I am rank number ", my_rank, " and my value of a is ", a
   call mpi_finalize(ierr)
end program my_prog
